src/coreclr/jit/codegenxarch.cpp

```diff
@@ -1148,34 +1149,53 @@ void CodeGen::genSIMDSplitReturn(GenTree* src, ReturnTypeDesc* retTypeDesc)
     regNumber reg0  = retTypeDesc->GetABIReturnReg(0);
     regNumber reg1  = retTypeDesc->GetABIReturnReg(1);
 
+    assert((reg0 != REG_COUNT) && (reg1 != REG_COUNT) && (opReg != REG_COUNT));
+
+    const bool srcIsFloatReg = genIsValidFloatReg(opReg);
+    const bool dstIsFloatReg = genIsValidFloatReg(reg0);
+    assert(srcIsFloatReg);
+
+#ifdef TARGET_AMD64
+    assert(src->TypeIs(TYP_SIMD16));
+    assert(srcIsFloatReg == dstIsFloatReg);
     if (opReg != reg0 && opReg != reg1)
     {
         // Operand reg is different from return regs.
         // Copy opReg to reg0 and let it to be handled by one of the
         // two cases below.
-        inst_RV_RV(ins_Copy(TYP_DOUBLE), reg0, opReg, TYP_DOUBLE);
+        inst_RV_RV(ins_Copy(opReg, TYP_SIMD16), reg0, opReg, TYP_SIMD16);
         opReg = reg0;
     }
 
     if (opReg == reg0)
     {
         assert(opReg != reg1);
-
-        // reg0 - already has required 8-byte in bit position [63:0].
         // reg1 = opReg.
-        // swap upper and lower 8-bytes of reg1 so that desired 8-byte is in bit position [63:0].
-        inst_RV_RV(ins_Copy(TYP_DOUBLE), reg1, opReg, TYP_DOUBLE);
+        inst_RV_RV(ins_Copy(opReg, TYP_SIMD16), reg1, opReg, TYP_SIMD16);
     }
     else
     {
         assert(opReg == reg1);
 
         // reg0 = opReg.
-        // swap upper and lower 8-bytes of reg1 so that desired 8-byte is in bit position [63:0].
-        inst_RV_RV(ins_Copy(TYP_DOUBLE), reg0, opReg, TYP_DOUBLE);
+
+        inst_RV_RV(ins_Copy(opReg, TYP_SIMD16), reg0, opReg, TYP_SIMD16);
     }
+    // reg0 - already has required 8-byte in bit position [63:0].
+    // swap upper and lower 8-bytes of reg1 so that desired 8-byte is in bit position [63:0].
     inst_RV_RV_IV(INS_shufpd, EA_16BYTE, reg1, reg1, 0x01);
+
+#else  // TARGET_X86
+    assert(src->TypeIs(TYP_SIMD8));
+    assert(srcIsFloatReg != dstIsFloatReg);
+    assert((reg0 == REG_EAX) && (reg1 == REG_EDX));
+    // reg0 = opReg[31:0]
+    inst_RV_RV(ins_Copy(opReg, TYP_INT), reg0, opReg, TYP_INT);
+    // reg1 = opRef[63:32]
+    inst_RV_RV_IV(INS_pextrd, EA_4BYTE, reg1, opReg, 1);
```

Comment:
Are we guaranteed to be able to use an SSE4.1 instruction here?

Reply:
If not, we should be able to use `PEXTRW` twice along with a 16 bit shift to emulate it.

Reply:
I can't find 5.0 requirements for some reason,  @echesakovMSFT or @tannergooding probably know better.

Reply:
In case SSE2 is our minimum, here's an instruction sequence that I think would be SSE2-compatible with equivalent results (given the return value is in xmm0) to save you time if needed:

```
pextrw  eax, xmm0, 3
shl     eax, 16
pextrw  edx, xmm0, 2
and     edx, eax
movd    eax, xmm0
```

Reply:
Yes, I agree with @jkoritzinsky - you would need to check for `compExactlyDependsOn(InstructionSet_SSE41)` before emitting `pextrw` in a similar way as it is done here https://github.com/dotnet/runtime/blob/08c3065756cc677b8478fa05cd493722e3f03a07/src/coreclr/jit/hwintrinsicxarch.cpp#L1218

Reply:
> In case SSE2 is our minimum, here's an instruction sequence that I think would be SSE2-compatible with equivalent results (given the return value is in xmm0) to save you time if needed:

Thanks @jkoritzinsky , probably we could do with fewer instructions:
```
movd eax, xmm0
pshufd xmm0, xmm0, 1
movd edx, xmm0
```
but I doubt that people would care about performance or code size while running on such machines without SSE4.

Reply:
I was looking for a shuffle instruction but I couldn't find it! Your snippet is much better

Reply:
I wonder if the following SSE2 version is slightly faster
```asm
movd    eax, xmm0
pshufd  xmm0, xmm0, 1
movd    edx, xmm0
```

Reply:
Just realized that Sergey posted the same 20 min ago :)

Reply:
> In addition, please make sure that we don't emit that instruction during crossgen.

This case happens only with CoreRun for a special interop stub so can't check it easily, but if `compExactlyDependsOn(InstructionSet_SSE41)` returns `true` in crossgen we would probably have many other tests catching this, would not we?

Reply:
> This case happens only with CoreRun for a special interop stub so can't check it easily, but if compExactlyDependsOn(InstructionSet_SSE41) returns true in crossgen we would probably have many other tests catching this, would not we?

@sandreenko I can't tell for sure. For intrinsics it is never an issue since they are guarded by `IsSupported`. But I remember there was a [discussion](https://github.com/dotnet/runtime/pull/38060#discussion_r443062650 ) where @davidwrighton mentioned that 

> We compile System.Private.CoreLib with the instruction set such that it thinks its compiling with support for SSE4.1 and such

and I concluded that it might return `true` in crossgen scenario.

Reply:
> and I concluded that it might return true in crossgen scenario.

I ran crossgen SPC with `assert(!compExactlyDependsOn(InstructionSet_SSE41));` and, of course, it failed, what is the best solution here, always generate SSE2 code or wrap it in many `FEATURE_READYTORUN_COMPILER` defines?

Reply:
I believe you want `compOpportunisticallyDependsOn`. The SSE2 vs SSE4.1 paths are equivalent, it's just that the SSE4.1 path is "faster" and should be preferred when it is available.

For crossgen1 `SSE4.1` should be false, for crossgen2, it should depend on the ISA configuration (which I believe currently defaults to SSE4.1) and should flag the method as "uses SSE4.1" in which case it will be rejitted if false.

IIRC, there are some special considerations when it's all internal and therefore a `IsSupported` check doesn't exist, but @jkotas and @davidwrighton are the two most familiar with this (David wrote the original logic and Jan made some recent changes in the area).

Reply:
> I believe you want `compOpportunisticallyDependsOn`

+1

Reply:
`compOpportunisticallyDependsOn` is what you want and it will do exactly the right thing when using crossgen2, but I believe it is not safe for this kind of purpose when compiling S.P.C with crossgen1. As we haven't quite gotten rid of crossgen1, you'll need to find some way to not generate the SSE4.1 support when crossgening S.P.C with original crossgen, or at least, wait until #47019 is in and enabled for both X86 and X64.

Reply:
I would prefer to merge this PR as is with unsafe behavior for x86 old machines because:
1. this code is emitted only for PInvoke stubs that we don't crossgen, so no real issue here;
2. it could only affect old machines;
3. it will automatically be fixed with a full switch to crossgen2;

the alternative is to merge the change with SSE2 and open an issue to return SSE4.1 code after crossgen1 deprecation.

Reply:
Good news. Crossgen2 was put in place to replace compilation for S.P.C over the weekend. The only time that S.P.C should now be compiled with crossgen1 is in self-contained app build scenarios before we transition those over as well, and as you state there is no actual use of this feature in S.P.C.

In light of that, I believe this is ok to check in.